File: sdks/typescript/src/queue.ts
/**
 * Synap TypeScript SDK - Queue System Module
 * 
 * Message queue operations with ACK/NACK support.
 * Includes reactive consumer patterns using RxJS.
 */

import { Observable, Subject, timer, EMPTY, defer } from 'rxjs';
import { 
  switchMap, 
  retry, 
  catchError, 
  filter, 
  takeUntil,
  share,
  mergeMap
} from 'rxjs/operators';
import type { SynapClient } from './client';
import type {
  QueueConfig,
  QueueMessage,
  PublishOptions,
  QueueStats,
  QueueConsumerOptions,
  ProcessedMessage,
} from './types';

/**
 * Queue System client with reactive support
 */
export class QueueManager {
  private stopSignals = new Map<string, Subject<void>>();

  constructor(private client: SynapClient) {}

  /**
   * Create a new queue
   */
  async createQueue(name: string, config?: QueueConfig): Promise<boolean> {
    const result = await this.client.sendCommand<{ success: boolean }>('queue.create', {
      name,
      config: config || {},
    });
    return result.success;
  }

  /**
   * Publish a message to a queue
   */
  async publish(
    queueName: string,
    payload: Uint8Array | string,
    options?: PublishOptions
  ): Promise<string> {
    // Convert payload to array of numbers for JSON serialization
    const payloadBytes =
      typeof payload === 'string'
        ? Array.from(new TextEncoder().encode(payload))
        : Array.from(payload);

    const cmdPayload: Record<string, any> = {
      queue: queueName,
      payload: payloadBytes,
    };

    if (options?.priority !== undefined) {
      cmdPayload.priority = options.priority;
    }

    if (options?.max_retries !== undefined) {
      cmdPayload.max_retries = options.max_retries;
    }

    if (options?.headers) {
      cmdPayload.headers = options.headers;
    }

    const result = await this.client.sendCommand<{ message_id: string }>(
      'queue.publish',
      cmdPayload
    );

    return result.message_id;
  }

  /**
   * Consume a message from a queue
   */
  async consume(queueName: string, consumerId: string): Promise<QueueMessage | null> {
    const result = await this.client.sendCommand<{ message: QueueMessage | null }>(
      'queue.consume',
      {
        queue: queueName,
        consumer_id: consumerId,
      }
    );

    if (!result.message) {
      return null;
    }

    // Convert payload array back to Uint8Array
    const message = result.message;
    if (Array.isArray(message.payload)) {
      message.payload = new Uint8Array(message.payload);
    }

    return message;
  }

  /**
   * Acknowledge message processing (ACK)
   */
  async ack(queueName: string, messageId: string): Promise<boolean> {
    const result = await this.client.sendCommand<{ success: boolean }>('queue.ack', {
      queue: queueName,
      message_id: messageId,
    });
    return result.success;
  }

  /**
   * Negative acknowledge (NACK) - requeue or send to DLQ
   */
  async nack(queueName: string, messageId: string, requeue: boolean = true): Promise<boolean> {
    const result = await this.client.sendCommand<{ success: boolean }>('queue.nack', {
      queue: queueName,
      message_id: messageId,
      requeue,
    });
    return result.success;
  }

  /**
   * Get queue statistics
   */
  async stats(queueName: string): Promise<QueueStats> {
    return this.client.sendCommand<QueueStats>('queue.stats', {
      queue: queueName,
    });
  }

  /**
   * List all queues
   */
  async listQueues(): Promise<string[]> {
    const result = await this.client.sendCommand<{ queues: string[] }>('queue.list', {});
    return result.queues;
  }

  /**
   * Purge all messages from a queue
   */
  async purge(queueName: string): Promise<number> {
    const result = await this.client.sendCommand<{ purged: number }>('queue.purge', {
      queue: queueName,
    });
    return result.purged;
  }

  /**
   * Delete a queue
   */
  async deleteQueue(queueName: string): Promise<boolean> {
    const result = await this.client.sendCommand<{ deleted: boolean }>('queue.delete', {
      queue: queueName,
    });
    return result.deleted;
  }

  /**
   * Helper: Publish a string message
   */
  async publishString(
    queueName: string,
    message: string,
    options?: PublishOptions
  ): Promise<string> {
    return this.publish(queueName, message, options);
  }

  /**
   * Helper: Publish a JSON object
   */
  async publishJSON(
    queueName: string,
    data: any,
    options?: PublishOptions
  ): Promise<string> {
    const json = JSON.stringify(data);
    return this.publish(queueName, json, options);
  }

  /**
   * Helper: Consume and decode as string
   */
  async consumeString(queueName: string, consumerId: string): Promise<{
    message: QueueMessage | null;
    text: string | null;
  }> {
    const message = await this.consume(queueName, consumerId);
    
    if (!message) {
      return { message: null, text: null };
    }

    const text = new TextDecoder().decode(message.payload);
    return { message, text };
  }

  /**
   * Helper: Consume and decode as JSON
   */
  async consumeJSON<T = any>(queueName: string, consumerId: string): Promise<{
    message: QueueMessage | null;
    data: T | null;
  }> {
    const result = await this.consumeString(queueName, consumerId);
    
    if (!result.text) {
      return { message: null, data: null };
    }

    try {
      const data = JSON.parse(result.text) as T;
      return { message: result.message, data };
    } catch (error) {
      throw new Error(`Failed to parse JSON: ${error}`);
    }
  }

  // ==================== REACTIVE METHODS ====================

  /**
   * Create a reactive message consumer as an Observable
   * 
   * @example
   * ```typescript
   * synap.queue.observeMessages({
   *   queueName: 'tasks',
   *   consumerId: 'worker-1',
   *   pollingInterval: 500,
   *   concurrency: 5
   * }).subscribe({
   *   next: async (msg) => {
   *     await processMessage(msg.data);
   *     await msg.ack();
   *   },
   *   error: (err) => console.error('Error:', err)
   * });
   * ```
   */
  observeMessages<T = any>(options: QueueConsumerOptions): Observable<ProcessedMessage<T>> {
    const {
      queueName,
      consumerId,
      pollingInterval = 1000,
      concurrency = 1,
      requeueOnNack = true,
    } = options;

    const stopKey = `${queueName}:${consumerId}`;
    const stopSignal = new Subject<void>();
    this.stopSignals.set(stopKey, stopSignal);

    return timer(0, pollingInterval).pipe(
      takeUntil(stopSignal),
      mergeMap(
        async () => {
          try {
            const result = await this.consumeJSON<T>(queueName, consumerId);
            
            if (!result.message || !result.data) {
              return null;
            }

            const processedMessage: ProcessedMessage<T> = {
              message: result.message,
              data: result.data,
              ack: async () => {
                await this.ack(queueName, result.message!.id);
              },
              nack: async (requeue: boolean = requeueOnNack) => {
                await this.nack(queueName, result.message!.id, requeue);
              },
            };

            return processedMessage;
          } catch (error) {
            console.error(`Error consuming from ${queueName}:`, error);
            return null;
          }
        },
        concurrency
      ),
      filter((msg): msg is ProcessedMessage<T> => msg !== null),
      share()
    ) as Observable<ProcessedMessage<T>>;
  }

  /**
   * Create a reactive message consumer with automatic ACK/NACK handling
   * 
   * @example
   * ```typescript
   * synap.queue.observeMessagesAuto({
   *   queueName: 'tasks',
   *   consumerId: 'worker-1',
   * }).subscribe({
   *   next: async (msg) => {
   *     // Process message - will auto-ACK on success
   *     await processMessage(msg.data);
   *   },
   *   error: (err) => console.error('Error:', err)
   * });
   * ```
   */
  observeMessagesAuto<T = any>(options: QueueConsumerOptions): Observable<ProcessedMessage<T>> {
    const opts = {
      ...options,
      autoAck: true,
      autoNack: true,
    };

    return this.observeMessages<T>(opts);
  }

  /**
   * Create a reactive consumer that processes messages with a handler function
   * 
   * @example
   * ```typescript
   * const subscription = synap.queue.processMessages({
   *   queueName: 'emails',
   *   consumerId: 'email-worker',
   *   concurrency: 10
   * }, async (data) => {
   *   await sendEmail(data);
   * }).subscribe({
   *   next: (result) => console.log('Processed:', result),
   *   error: (err) => console.error('Error:', err)
   * });
   * ```
   */
  processMessages<T = any>(
    options: QueueConsumerOptions,
    handler: (data: T, message: QueueMessage) => Promise<void>
  ): Observable<{ messageId: string; success: boolean; error?: Error }> {
    return this.observeMessages<T>(options).pipe(
      mergeMap(
        async (msg) => {
          try {
            await handler(msg.data, msg.message);
            await msg.ack();
            return { messageId: msg.message.id, success: true };
          } catch (error) {
            await msg.nack();
            return { 
              messageId: msg.message.id, 
              success: false, 
              error: error as Error 
            };
          }
        },
        options.concurrency || 1
      )
    );
  }

  /**
   * Stop a reactive consumer
   * 
   * @param queueName - Queue name
   * @param consumerId - Consumer ID
   */
  stopConsumer(queueName: string, consumerId: string): void {
    const stopKey = `${queueName}:${consumerId}`;
    const stopSignal = this.stopSignals.get(stopKey);
    
    if (stopSignal) {
      stopSignal.next();
      stopSignal.complete();
      this.stopSignals.delete(stopKey);
    }
  }

  /**
   * Stop all reactive consumers
   */
  stopAllConsumers(): void {
    this.stopSignals.forEach((signal) => {
      signal.next();
      signal.complete();
    });
    this.stopSignals.clear();
  }

  /**
   * Create an observable that emits queue statistics at regular intervals
   * 
   * @param queueName - Queue name
   * @param interval - Polling interval in milliseconds (default: 5000)
   * 
   * @example
   * ```typescript
   * synap.queue.observeStats('tasks', 1000).subscribe({
   *   next: (stats) => console.log('Queue depth:', stats.depth),
   * });
   * ```
   */
  observeStats(queueName: string, interval: number = 5000): Observable<QueueStats> {
    return timer(0, interval).pipe(
      switchMap(() => defer(() => this.stats(queueName))),
      retry({
        count: 3,
        delay: 1000,
      }),
      catchError((error: unknown) => {
        console.error(`Error fetching stats for ${queueName}:`, error);
        return EMPTY;
      }),
      share()
    );
  }
}

File: sdks/rust/src/queue.rs
//! Queue operations

use crate::client::SynapClient;
use crate::error::{Result, SynapError};
use crate::types::{Message, QueueStats};
use serde_json::json;

// Re-export for convenience
pub use crate::reactive::{MessageStream, SubscriptionHandle};

/// Queue Manager interface
#[derive(Clone)]
pub struct QueueManager {
    pub(crate) client: SynapClient,
}

impl QueueManager {
    /// Create a new Queue manager interface
    pub(crate) fn new(client: SynapClient) -> Self {
        Self { client }
    }

    /// Create a new queue
    ///
    /// # Arguments
    /// * `queue_name` - Name of the queue
    /// * `max_depth` - Maximum queue depth (optional)
    /// * `ack_deadline_secs` - ACK deadline in seconds (optional)
    pub async fn create_queue(
        &self,
        queue_name: &str,
        max_depth: Option<usize>,
        ack_deadline_secs: Option<u64>,
    ) -> Result<()> {
        let mut body = json!({});
        if let Some(depth) = max_depth {
            body["max_depth"] = json!(depth);
        }
        if let Some(deadline) = ack_deadline_secs {
            body["ack_deadline_secs"] = json!(deadline);
        }

        let path = format!("queue/{}", queue_name);
        let response = self.client.post(&path, body).await?;

        if response["success"].as_bool() != Some(true) {
            return Err(SynapError::ServerError(
                "Failed to create queue".to_string(),
            ));
        }

        Ok(())
    }

    /// Publish a message to a queue
    ///
    /// # Example
    /// ```no_run
    /// # use synap_sdk::{SynapClient, SynapConfig};
    /// # #[tokio::main]
    /// # async fn main() -> Result<(), Box<dyn std::error::Error>> {
    /// # let client = SynapClient::new(SynapConfig::new("http://localhost:15500"))?;
    /// client.queue().publish("tasks", b"process-video", Some(9), None).await?;
    /// # Ok(())
    /// # }
    /// ```
    pub async fn publish(
        &self,
        queue_name: &str,
        payload: &[u8],
        priority: Option<u8>,
        max_retries: Option<u32>,
    ) -> Result<String> {
        let path = format!("queue/{}/publish", queue_name);
        let body = json!({
            "payload": payload,
            "priority": priority,
            "max_retries": max_retries,
        });

        let response = self.client.post(&path, body).await?;

        Ok(response["message_id"]
            .as_str()
            .unwrap_or_default()
            .to_string())
    }

    /// Consume a message from a queue
    pub async fn consume(&self, queue_name: &str, consumer_id: &str) -> Result<Option<Message>> {
        let path = format!("queue/{}/consume/{}", queue_name, consumer_id);
        let response = self.client.get(&path).await?;

        if response.is_null() {
            return Ok(None);
        }

        Ok(serde_json::from_value(response).ok())
    }

    /// Acknowledge a message
    pub async fn ack(&self, queue_name: &str, message_id: &str) -> Result<()> {
        let path = format!("queue/{}/ack", queue_name);
        let body = json!({"message_id": message_id});

        let response = self.client.post(&path, body).await?;

        if response["success"].as_bool() != Some(true) {
            return Err(SynapError::ServerError("Failed to ACK message".to_string()));
        }

        Ok(())
    }

    /// Negative acknowledge a message (requeue)
    pub async fn nack(&self, queue_name: &str, message_id: &str) -> Result<()> {
        let path = format!("queue/{}/nack", queue_name);
        let body = json!({"message_id": message_id});

        let response = self.client.post(&path, body).await?;

        if response["success"].as_bool() != Some(true) {
            return Err(SynapError::ServerError(
                "Failed to NACK message".to_string(),
            ));
        }

        Ok(())
    }

    /// Get queue statistics
    pub async fn stats(&self, queue_name: &str) -> Result<QueueStats> {
        let path = format!("queue/{}/stats", queue_name);
        let response = self.client.get(&path).await?;
        Ok(serde_json::from_value(response)?)
    }

    /// List all queues
    pub async fn list(&self) -> Result<Vec<String>> {
        let response = self.client.get("queue/list").await?;
        Ok(serde_json::from_value(response["queues"].clone())?)
    }

    /// Delete a queue
    pub async fn delete_queue(&self, queue_name: &str) -> Result<()> {
        let path = format!("queue/{}", queue_name);
        let response = self.client.delete(&path).await?;

        if response["success"].as_bool() != Some(true) {
            return Err(SynapError::ServerError(
                "Failed to delete queue".to_string(),
            ));
        }

        Ok(())
    }
}
